spec/client/ics-008-wasm-client/README.md

```diff
@@ -84,329 +111,450 @@ interface Height {
+  pub timestamp: Option<u64>,
 }
 
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
-#[serde(rename_all = "snake_case")]
-pub struct CreateConsensusMessage {
-    pub client_state: Vec<u8>,
-    pub height: Height
+#[cw_serde]
+pub struct GenesisMetadata {
+  pub key: Vec<u8>,
+  pub value: Vec<u8>,
 }
+```
 
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
-#[serde(rename_all = "snake_case")]
-pub struct InitializeClientStateMessage {
-    pub initialization_data: Vec<u8>,
-    pub consensus_state: Vec<u8>
-}
+#### Contract execute
 
+Every Wasm contract must support these execute messages:
 
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
-#[serde(rename_all = "snake_case")]
-pub enum HandleMsg {
-    HandleMisbehaviour(MisbehaviourMessage),
-    TryCreateConsensusState(CreateConsensusMessage),
-    InitializeClientState(InitializeClientStateMessage)
+```rust
+#[cw_serde]
+pub struct UpdateStateOnMisbehaviourMsg {
+  pub client_message: ClientMessage,
 }
 
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
-#[serde(rename_all = "snake_case")]
-pub struct ValidateClientStateMessage {
-    client_state: Vec<u8>,
-    height: Height
+#[cw_serde]
+pub struct UpdateStateMsg {
+  pub client_message: ClientMessage,
 }
 
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
-#[serde(rename_all = "snake_case")]
-pub struct ValidateNewClientStateMessage {
-    client_state: Vec<u8>,
-    new_client_state: Vec<u8>,
-    height: Height
+#[cw_serde]
+pub struct CheckSubstituteAndUpdateStateMsg {
+  substitute_client_msg: Vec<u8>,
 }
 
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
-#[serde(rename_all = "snake_case")]
-pub struct ValidateInitializationDataMessage {
-    init_data: Vec<u8>,
-    consensus_state: Vec<u8>
+#[cw_serde]
+pub struct VerifyUpgradeAndUpdateStateMsgRaw {
+  pub upgrade_client_state: WasmClientState<FakeInner, FakeInner, FakeInner>,
+  pub upgrade_consensus_state: WasmConsensusState<FakeInner>,
+  #[schemars(with = "String")]
+  #[serde(with = "Base64", default)]
+  pub proof_upgrade_client: Vec<u8>,
+  #[schemars(with = "String")]
+  #[serde(with = "Base64", default)]
+  pub proof_upgrade_consensus_state: Vec<u8>,
 }
+```
 
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
-#[serde(rename_all = "snake_case")]
-pub enum ValidityPredicate {
-    ClientState(ValidateClientStateMessage),
-    NewClientState(ValidateNewClientStateMessage),
-    InitializationData(ValidateInitializationDataMessage),
-}
+The response for execute is as follows:
 
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
-#[serde(rename_all = "snake_case")]
-pub enum QueryMsg {
-   IsValid(ValidityPredicate),
-   LatestClientHeight(Vec<u8>),
+```rust
+#[cw_serde]
+pub struct ContractResult {
+  pub is_valid: bool, // Execute was successful
+  pub error_msg: String, // Error message
```

Comment:
this was brought up in slack discussion at some point, I remember both `is_valid` and `error_msg` being unused in contracts. Either way, can be left for another PR after we actually remove em from code.